ggml-opencl.c

```diff
@@ -188,48 +186,143 @@ static cl_program build_program_from_source(cl_context ctx, cl_device_id dev, co
+    if (GGML_OPENCL_PLATFORM != NULL) {
+        unsigned plat_num;
+        if (sscanf(GGML_OPENCL_PLATFORM, " %u", &plat_num) == 1) {
+            if (plat_num >= num_platforms) {
+                fprintf(stderr, "ggml_opencl: There is no platform %d\n", plat_num);
+                exit(1);
+            } else {
+                platform = platforms[plat_num];
+                CL_CHECK(clGetPlatformInfo(platform, CL_PLATFORM_NAME, sizeof(text_buffer), &text_buffer, NULL));
+            }
+        } else {
+            for (unsigned i = 0; i < num_platforms; i++) {
+                CL_CHECK(clGetPlatformInfo(platforms[i], CL_PLATFORM_NAME, sizeof(text_buffer), &text_buffer, NULL));
+                if (strstr(text_buffer, GGML_OPENCL_PLATFORM) != NULL) {
+                    platform = platforms[i];
+                    break;
+                }
+            }
+        }
+        if (platform == NULL) {
+            fprintf(stderr, "ggml_opencl: no platform matching '%s' was found.\n", GGML_OPENCL_PLATFORM);
+            exit(1);
+        } else {
+            fprintf(stderr, "ggml_opencl: selecting platform: '%s'\n", text_buffer);
+        }
+    }
+
+    char * GGML_OPENCL_DEVICE = getenv("GGML_OPENCL_DEVICE");
+    if (GGML_OPENCL_DEVICE != NULL) {
+        cl_device_id devices[NDEV];
+        cl_uint num_devices;
+        CL_CHECK(clGetDeviceIDs(platform, CL_DEVICE_TYPE_ALL, NDEV, devices, &num_devices));
+
+        unsigned dev_num;
+        if (sscanf(GGML_OPENCL_DEVICE, " %u", &dev_num) == 1) {
+            if (dev_num >= num_devices) {
+                fprintf(stderr, "ggml_opencl: There is no device %d\n", dev_num);
+                exit(1);
+            } else {
+                device = devices[dev_num];
+                CL_CHECK(clGetDeviceInfo(device, CL_DEVICE_NAME, sizeof(text_buffer), &text_buffer, NULL));
+            }
+        } else {
+            for (unsigned i = 0; i < num_devices; i++) {
+                CL_CHECK(clGetDeviceInfo(devices[i], CL_DEVICE_NAME, sizeof(text_buffer), &text_buffer, NULL));
+                if (strstr(text_buffer, GGML_OPENCL_DEVICE) != NULL) {
+                    device = devices[i];
+                    break;
+                }
+            }
+        }
+        if (device == NULL) {
+            fprintf(stderr, "ggml_opencl: no device matching '%s' was found.\n", GGML_OPENCL_DEVICE);
+            exit(1);
+        } else {
+            fprintf(stderr, "ggml_opencl: selecting device: '%s'\n", text_buffer);
+            if (platform == NULL) {
+                CL_CHECK(clGetDeviceInfo(device, CL_DEVICE_PLATFORM, sizeof(&platform), &platform, NULL));
+            }
+        }
+    }
+
+    if (platform == NULL) {
+        cl_device_id devices[NDEV];
+        cl_uint num_devices;
+
+        for (unsigned i = 0; i < num_platforms; i++) {
+            CL_CHECK(clGetDeviceIDs(platforms[i], CL_DEVICE_TYPE_GPU, NDEV, devices, &num_devices));
```

Comment:
My gaming PC has POCL installed (CPU OpenCL) and fails here if I don't select devices, because the first platform (POCL) has no GPU device.
`ggml_opencl: clGetDeviceIDs(platforms[i], CL_DEVICE_TYPE_GPU, NDEV, devices, &num_devices) error -1 at ggml-opencl.c:269`

Reply:
Right, I will fix it.